ext/opentelemetry-ext-wsgi/src/opentelemetry/ext/wsgi/__init__.py

```diff
@@ -0,0 +1,105 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+The opentelemetry-ext-wsgi package provides a WSGI middleware that can be used
+on any WSGI framework (such as Django / Flask) to track requests timing through
+OpenTelemetry.
+"""
+
+import functools
+import wsgiref.util as wsgiref_util
+
+from opentelemetry import trace
+from opentelemetry.ext.wsgi.version import __version__  # noqa
+
+
+class OpenTelemetryMiddleware:
+    """The WSGI application middleware.
+
+    This class is used to create and annotate spans for requests to a WSGI
+    application.
+
+    Args:
+        wsgi: The WSGI application callable.
+    """
+
+    def __init__(
+            self,
+            wsgi,
+            span_context_propagator=None,
+            distributed_context_propagator=None,
+    ):
+        self.wsgi = wsgi
+
+        # TODO: implement context propagation
+        self.span_context_propagator = span_context_propagator
+        self.distributed_context_propagator = distributed_context_propagator
+
+    @staticmethod
+    def _add_request_attributes(span, environ):
+        span.set_attribute("component", "http")
+        span.set_attribute("http.method", environ["REQUEST_METHOD"])
+
+        host = environ.get("HTTP_HOST") or environ["SERVER_NAME"]
+        span.set_attribute("http.host", host)
+
+        url = (
+            environ.get("REQUEST_URI") or
+            environ.get("RAW_URI") or
+            wsgiref_util.request_uri(environ, include_query=False)
+        )
+        span.set_attribute("http.url", url)
+
+    @staticmethod
+    def _add_response_attributes(span, status):
+        status_code, status_text = status.split(" ", 1)
+        span.set_attribute("http.status_text", status_text)
+
+        try:
+            status_code = int(status_code)
+        except ValueError:
+            pass
+        else:
+            span.set_attribute("http.status_code", status_code)
+
+    @classmethod
+    def _create_start_response(cls, span, start_response):
+        @functools.wraps(start_response)
+        def _start_response(status, response_headers, *args):
+            cls._add_response_attributes(span, status)
+            return start_response(status, response_headers, *args)
+
+        return _start_response
+
+    def __call__(self, environ, start_response):
+        """The WSGI application
+
+        Args:
+            environ: A WSGI environment.
+            start_response: The WSGI start_response callable.
+        """
+
+        tracer = trace.tracer()
```

Comment:
We might need to explore how to specify custom tracer for integrations in a separate PR.

Reply:
Ah cool, thanks for pointing that out. I thought that was handled internally by the API loader. Is the tracer not configured through [set_preferred_tracer_implementation](https://github.com/open-telemetry/opentelemetry-python/blob/b7b38a5d5477797553ff0b85a690f55c9436e009/opentelemetry-api/src/opentelemetry/trace/__init__.py#L401)?

Reply:
I understand this differently, I could be wrong:
1. `set_preferred_tracer_implementation` is used to specify the SDK (which implementation to use).
2. Here the question is around which tracer **instance** to be used. Even with one SDK, the user could choose to have multiple tracers (similar like loggers in Python).

Reply:
The tracer [here](https://github.com/open-telemetry/opentelemetry-python/blob/b7b38a5d5477797553ff0b85a690f55c9436e009/opentelemetry-api/src/opentelemetry/trace/__init__.py#L386L398) is a singleton though. To create multiple tracer instances wouldn't you have to bypass the OpenTelemetry `opentelemetry.tracer()` API?

Reply:
The topic is beyond this PR.
https://github.com/open-telemetry/rfcs/pull/16/files is a better place to continue the discussion.